### PR TITLE
Update various libraries

### LINKED ISSUE #2411: [TASK] update libraries so Python 3.11 is usable
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
There are several libraries in RAVEN that need to be updated for Python 3.11 to be supported.

**Describe alternatives you've considered**
Continue using older libraries and don't support Python 3.11

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes #2411 

##### What are the significant changes in functionality due to this change request?
Besides the changes that are also in #2250
Drop labelEncoder when pickling KerasBase to support tensorflow 2.14
Convert to tuple before using np.hstack to support numpy 1.26
Explicitly use .squeeze() when needed for to support xarray 2024.7

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: ravenframework/SupervisedLearning/ARMA.py [hunk 1]
```diff
@@ -1628,7 +1628,10 @@ def getFundamentalFeatures(self, requestedFeatures, featureTemplate=None):
           ## IND
           #most probabble index
           if len(group['Ind']):
-            modeInd = stats.mode(group['Ind'])[0][0]
+            try:
```
Comment (wangcj05) [inline ravenframework/SupervisedLearning/ARMA.py:1631]: Could you provide a comment here to explain why we need the try except here?

File: tests/framework/ROM/SKLearn/gold/data/outOVO.csv [hunk 2]
```diff
@@ -1,122 +1,122 @@
...
-2.0,-1000.0,1,0.0025
-2.0,-800.0,1,0.005
-2.0,-600.0,1,0.005
-2.0,-400.0,1,0.005
-2.0,-200.0,1,0.005
-2.0,0.0,1,0.005
-2.0,200.0,1,0.005
-2.0,400.0,1,0.005
-2.0,600.0,1,0.005
+2.0,-1000.0,0,0.0025
```
Comment (wangcj05) [inline tests/framework/ROM/SKLearn/gold/data/outOVO.csv:2]: Could you provide a comment why all values of z are zeros instead of ones?

**Review by wangcj05 — CHANGES_REQUESTED** [review]
Some minor comments for you to consider. In addition, could you provide some explanation for the regolding tests? Some are minor changes which I assume due to the library updates, some changes may indicate a problem. Please comment on each regold test.

File: ravenframework/SupervisedLearning/ARMA.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply ravenframework/SupervisedLearning/ARMA.py:1631]: Hm, depending on version of Scipy, we get a different shaped array because of the change of the keepdims in stats.mode https://docs.scipy.org/doc/scipy/release/1.11.0-notes.html#expired-deprecations

So probably we should just specify keepdims here and use only one of these options.

File: tests/framework/ROM/SKLearn/gold/data/outOVO.csv (hunk 2 above)
Comment (joshua-cogliati-inl) [reply tests/framework/ROM/SKLearn/gold/data/outOVO.csv:2]: I think because the OVO classifier is using:
```
    <ROM name="estimator" subType="LinearRegression">
      <Features>X,Y</Features>
      <Target>Z</Target>
      <fit_intercept>True</fit_intercept>
      <normalize>True</normalize>
    </ROM>
``` 
underneath, and a linear regression is not a good fit for modeling this.
As I mentioned in the commit, this is because of this change: https://github.com/scikit-learn/scikit-learn/pull/22604

**Comment by joshua-cogliati-inl** [comment]
> Some minor comments for you to consider. In addition, could you provide some explanation for the regolding tests? Some are minor changes which I assume due to the library updates, some changes may indicate a problem. Please comment on each regold test.

* eb93d448411676b189f4f18d2f7b6e2a6d0f6d46 Regolds tests/framework/PostProcessors/Metrics/gold/scipyMetricsBoolean/pp1_print.csv due to pending removal of kulsinski metric.
* e3955f21ce850a1806c5e7bbdb6c852aa9c0b0aa Regolds tests/framework/ROM/SKLearn/gold/data/outOVO.csv due to fix in scipy for OVO https://github.com/scikit-learn/scikit-learn/pull/22604
* e352a568d0226bac3e464c396eae70bff7095124 Regolds several files in tests/framework/ROM/TimeSeries/PolyExponential/gold/ because of scipy changes. (These were rel err of 1e-03 or less, so the results are generally the same to 3 significant figures)
* 200c8f45183b109ec115b25761b3df4d0502a368 Regolds tests/framework/PostProcessors/Validation/gold/DSS/pp1_print_0.csv and tests/framework/PostProcessors/Validation/gold/DSS/pp3_print_0.csv. I am guessing this is from changes between scipy.integrate.simps and scipy.integrate.simpson
* 938e8bee0168f65afeff8c3bbb64381add14a377. Forces data to have uniform timestep spacing because HAVOK DMD now requires uniform spacing, which changes the results and resulted in regolding:   tests/framework/ROM/TimeSeries/DMD/gold/HAVOK/outputHAVOKDMD/outputDMD_0.csv
* 1851ed2b5d003def998065a00d59e11a8b1630f7 Previously we checked Fourier__signal_f__period10.0__phase, but it randomly is either +pi or -pi depending on OS, library version and other minor changes. This regolds tests/framework/PostProcessors/TSACharacterizer/gold/Basic/chz.csv and tests/framework/PostProcessors/TSACharacterizer/gold/Basic/windowsChz.csv to remove Fourier__signal_f__period10.0__phase.
* 954436296391f135647fb0ef78262b4249ed4ff6 Regolds tests/framework/ROM/TimeSeries/DMD/gold/BOPDMD/outputBOPDMD/outputDMD_0.csv because of library changes. These are similar to 1 significant figure.

**Review by wangcj05 — APPROVED** [review]
changes are good for me.

**Comment by wangcj05** [comment]
@joshua-cogliati-inl This PR is good for me. I have two questions:
1. Have you tested the parallel on our HPC using this PR?
2. When do you want to merge it?

File: ravenframework/SupervisedLearning/ARMA.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply ravenframework/SupervisedLearning/ARMA.py]: I switched it to use an explicit keepdims.

File: tests/framework/ROM/SKLearn/gold/data/outOVO.csv (hunk 2 above)
Comment (joshua-cogliati-inl) [reply tests/framework/ROM/SKLearn/gold/data/outOVO.csv]: I changed the estimator to Gaussian naive Bayes and the values are now a mixture of zeros and ones.

**Review by wangcj05 — APPROVED** [review]
changes are good.

**Review by wangcj05 — APPROVED** [review]
changes are good.

**Comment by wangcj05** [comment]
checklist is good, and PR can be merged.
